src/MessageFormatter.php

```diff
@@ -0,0 +1,180 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Log;
+
+use DateTime;
+
+use function gettype;
+use function implode;
+use function is_array;
+use function is_string;
+use function microtime;
+use function sprintf;
+use function strpos;
+
+final class MessageFormatter
+{
+    /**
+     * @var callable|null PHP callable that returns a string representation of the log message.
+     *
+     * If not set, {@see MessageFormatter::defaultFormat()} will be used.
+     *
+     * The signature of the callable should be `function (array $message): string;`.
+     */
+    private $format;
+
+    /**
+     * @var callable|null PHP callable that returns a string to be prefixed to every exported message.
+     *
+     * If not set, {@see MessageFormatter::getPrefix()} will be used, which prefixes
+     * the message with context information such as user IP, user ID and session ID.
+     *
+     * The signature of the callable should be `function (array $message): string;`.
+     */
+    private $prefix;
+
+    /**
+     * @var string The date format for the log timestamp. Defaults to `Y-m-d H:i:s.u`.
+     */
+    private string $timestampFormat = 'Y-m-d H:i:s.u';
+
+    /**
+     * Sets the format for the string representation of the log message.
+     *
+     * @param callable $format The PHP callable to get a string representation of the log message.
+     *
+     * @see MessageFormatter::$format
+     */
+    public function setFormat(callable $format): void
```

Comment:
Do we need the class to be mutable? i.e. do we need to change format in runtime?

Reply:
Yes, the class must be mutable, it will give the opportunity to change the format in the child target.